### PR TITLE
NfcDict Refactoring

### PR DESCRIPTION
**Pull request by v0lp3**
# What's new

I refactored the NfcDict structure to make it more generic,  so now we can use it with other applications like RFID (for example if we implement dictionary attack on T5577's password). The NfcDict was renamed KeysList and was moved into toolbox, therefore MfUserDict wraps this structure.

# Verification 

Test NFC application.

# Checklist (For Reviewer)

- [x] PR has description of feature/bug or link to Confluence/Jira task
- [x] Description contains actions to verify feature/bugfix
- [x] I've built this code, uploaded it to the device and verified feature/bugfix

### CODE REVIEW

File: targets/f7/api_symbols.csv [hunk 1]
```diff
@@ -1,5 +1,5 @@
 entry,status,name,type,params
-Version,+,49.1,,
+Version,v,50.0,,
```
Comment (hedger) [inline targets/f7/api_symbols.csv]: don't leave "v" here.

Build both targets locally first: `./fbt TARGET_HW=18 updater_package`

**Comment by skotopes** [comment]
@v0lp3 also keep in mind that there are other apps that using nfc dict (picopass) that is going to be broken by those changes

**Comment by v0lp3** [comment]
> @v0lp3 also keep in mind that there are other apps that using nfc dict (picopass) that is going to be broken by those changes

I see. How about we retain the nfc_dict.* files or replace functions in nfc_dict.h with inlines to ensure compatibility?

**Comment by skotopes** [comment]
@v0lp3 good question, let's wait for @gornekich comment on this PR.

File: applications/main/nfc/scenes/nfc_scene_config.h [hunk 2]
```diff
@@ -48,7 +48,7 @@ ADD_SCENE(nfc, mf_classic_write_initial_fail, MfClassicWriteInitialFail)
 ADD_SCENE(nfc, mf_classic_wrong_card, MfClassicWrongCard)
 
 ADD_SCENE(nfc, mf_classic_keys, MfClassicKeys)
-ADD_SCENE(nfc, mf_classic_keys_list, MfClassicKeysList)
```
Comment (gornekich) [inline applications/main/nfc/scenes/nfc_scene_config.h]: Undo this change

**Review by gornekich — CHANGES_REQUESTED** [review]
Please, rename KeysList to KeysDict and rework nfc application with KeysDict. No need to support old symbols. I will rework picopass app to use new KeysDict by my self.

So, the scope of this PR should be moving NfcDict to KeysDict in toolbox

**Comment by v0lp3** [comment]
> Please, rename KeysList to KeysDict and rework nfc application with KeysDict. No need to support old symbols. I will rework picopass app to use new KeysDict by my self.
> 
> So, the scope of this PR should be moving NfcDict to KeysDict in toolbox

Ok done

**Comment by gornekich** [comment]
@v0lp3 please, fix unit tests.
Run `./fbt FIRMWARE_APP_SET=unit_tests` to build firmware with unit tests. Then connect to cli via `./fbt cli`, run `unit_tests nfc` and make sure all tests are passed

**Comment by v0lp3** [comment]
> @v0lp3 please, fix unit tests. Run `./fbt FIRMWARE_APP_SET=unit_tests` to build firmware with unit tests. Then connect to cli via `./fbt cli`, run `unit_tests nf`c and make sure all tests are passed

How can i configure a mock flipper zero device?

**Comment by gornekich** [comment]
> > @v0lp3 please, fix unit tests. Run `./fbt FIRMWARE_APP_SET=unit_tests` to build firmware with unit tests. Then connect to cli via `./fbt cli`, run `unit_tests nf`c and make sure all tests are passed
> 
> How can i configure a mock flipper zero device?

Don't understand your problem. Do you have flipper to test your changes?

**Comment by v0lp3** [comment]
I was wondering if it is a problem to run unit tests with PR GitHub actions. I read that the device used in the actions is mock and I was wondering if it would be possible to replicate the configuration in my own environment

**Comment by v0lp3** [comment]
> > > @v0lp3 please, fix unit tests. Run `./fbt FIRMWARE_APP_SET=unit_tests` to build firmware with unit tests. Then connect to cli via `./fbt cli`, run `unit_tests nf`c and make sure all tests are passed
> > 
> > 
> > How can i configure a mock flipper zero device?
> 
> Don't understand your problem. Do you have flipper to test your changes?

I was wondering if it is a problem to run unit tests with PR GitHub actions. I read that the device used in the actions is mock and I was asking you if it would be possible to replicate the configuration in my own environment.

File: lib/nfc/protocols/mf_classic/mf_classic.h [hunk 3]
```diff
@@ -232,6 +240,48 @@ bool mf_classic_is_allowed_access(
     MfClassicKeyType key_type,
     MfClassicAction action);
 
+inline __attribute__((always_inline)) MfClassicDict*
+    mf_classic_dict_alloc(const char* path, MfClassicDictMode mode) {
+    return (MfClassicDict*)keys_dict_alloc(path, (KeysDictMode)mode, MF_CLASSIC_KEY_SIZE);
+}
```
Comment (gornekich) [inline lib/nfc/protocols/mf_classic/mf_classic.h]: Please, delete these mf_classic_dict* functions here and below. It's totally OK if nfc app will use KeysDict directly.

File: lib/nfc/protocols/mf_classic/mf_classic.h [hunk 4]
```diff
@@ -34,6 +35,13 @@ extern "C" {
 #define MF_CLASSIC_AR_SIZE (4)
 #define MF_CLASSIC_AT_SIZE (4)
 
+typedef KeysDict MfClassicDict;
+
+typedef enum {
+    MfClassicDictModeOpenExisting,
+    MfClassicDictModeOpenAlways,
+} MfClassicDictMode;
+
```
Comment (gornekich) [inline lib/nfc/protocols/mf_classic/mf_classic.h]: Unnecessary types

File: lib/toolbox/keys_dict.c [hunk 5]
```diff
@@ -0,0 +1,324 @@
...
+#define TAG "KeysDict"
+
+struct KeysDict {
+    Stream* stream;
+    size_t key_size;
+    size_t key_size_symbols;
+    size_t total_keys;
+};
+
+static inline void __add_ending_new_line(KeysDict* instance) {
```
Comment (gornekich) [inline lib/toolbox/keys_dict.c]: Please, avoid such naming. Even static functions should starts with lib name. This function should be names keys_dict_add_ending_new_line

File: lib/toolbox/keys_dict.c [hunk 6]
```diff
@@ -0,0 +1,324 @@
...
+        if(stream_read(instance->stream, &last_char, 1) == 1 && last_char != '\n') {
+            FURI_LOG_D(TAG, "Adding new line ending");
+            stream_write_char(instance->stream, '\n');
+        }
+
+        stream_rewind(instance->stream);
+    }
+}
+
+static inline bool __read_key_line(KeysDict* instance, FuriString* line, bool* is_endfile) {
```
Comment (gornekich) [inline lib/toolbox/keys_dict.c]: Same here. keys_dict_read_line()

File: lib/toolbox/keys_dict.c [hunk 7]
```diff
@@ -0,0 +1,324 @@
...
+    if(stream_read_line(instance->stream, line) == false)
+        *is_endfile = true;
+
+    else {
```
Comment (gornekich) [inline lib/toolbox/keys_dict.c]: Please, add here and below brackets {}. It will be much easier to read and allows to avoid mistakes in future.

**Review by gornekich — CHANGES_REQUESTED** [review]

**Comment by gornekich** [comment]
> I was wondering if it is a problem to run unit tests with PR GitHub actions. I read that the device used in the actions is mock and I was asking you if it would be possible to replicate the configuration in my own environment.

I see that you fixed unit tests. Running locally `unit_tests` in CLI is equivalent to CI tests

**Review by gornekich — APPROVED** [review]
Well done, thanks!

**Review by skotopes — APPROVED** [review]

**Pull request merged**